pkg/volume/flexvolume/unmounter.go

```diff
@@ -42,15 +42,20 @@ func (f *flexVolumeUnmounter) TearDown() error {
 }
 
 func (f *flexVolumeUnmounter) TearDownAt(dir string) error {
-
 	pathExists, pathErr := mount.PathExists(dir)
 	if !pathExists {
 		klog.Warningf("Warning: Unmount skipped because path does not exist: %v", dir)
 		return nil
 	}
 
-	if pathErr != nil && !mount.IsCorruptedMnt(pathErr) {
-		return fmt.Errorf("Error checking path: %v", pathErr)
+	if pathErr != nil {
+		// only log warning here since plugins should anyways have to deal with errors
+		klog.Warningf("Error checking path: %v", pathErr)
+	} else {
+		if !pathExists {
```

Comment:
we rely on the correctness of mount.PathExists(dir) to return not exist only if path is really no longer exist. but the current implementation seems still has problem. https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount_helper_common.go#L101
should we fix that too?

Reply:
hi @jingxu97 originally I fixed this issue like this:
https://github.com/andyzhangx/kubernetes/commit/f99d3ab223eda0a1c341b67b04aa968b83f8a2d6
Which way do you perfer?  also loop in @gnufied  @msau42

Reply:
I think your current fix plus the fix you had for mount.PathExists(dir)  is good enough.

Reply:
I think the main concern is if we fix PathExists, that impacts all plugins

Reply:
understand. But if os.IsNotExist(err) is not true, we are not sure the path exists or not. Currently plugins will return nil instead of unmounting if path is not exists, which is wrong. I think the change is  actually makes the system perform in a safer way.

Reply:
@jingxu97 @msau42 PTAL, I have changed the code of function `PathExits`, it will always return `<true, error>` if path has error

Reply:
I have looked at all the places of calling PathExists, there are two ways of checking
1. If PathErr != nil: First check whether there is error or not, if there is error, return the **error** immediately in the function. Otherwise, it checks !pathExists, if it is true (not exist), skip cleanup steps and immediately return. If it is false (exist), it will do cleanup step 
2. If !PathExists: First check whether the boolean value of PathExists is false or not. If it is false, normally skip some cleanup steps and return **nil** (no error)

With this the change of returning true instead of false in PathExists, I think it is safe to use in both checking approaches. 
For option 1, if PathErr != nil, it excludes the cases of PathExists=true without error or PathExists=false (basically we are sure of path either exists or not exists without any other checking errors). After this step, it continues to check PathExists or not and determine the action.
For option 2, if !PathExists is true, the new change we have here means that we are sure that path does not exist, and we can safely skip some clean up action. Before we have the problem of setting !PathExists is true even in some error cases that we are actually not sure about the existence.
In summary, I think it is safe to make this change.

Reply:
Why do we need to check pathExists again here? It's already been changed https://github.com/kubernetes/kubernetes/pull/75234/files#diff-e39e9fce3fb72bfbb3f128662eb6639dR46

I think this change is related, and may potentially be needed? https://github.com/kubernetes/kubernetes/pull/75645

Before, if there's some non-corrupted error, pathExists = false and the Unmount operation will get skipped and return success.  Now with this change, pathExists = true and pathErr != nil, TearDown returns early before calling Unmount, so now, Teardown returns error forever.

Reply:
@msau42 good catch, I have changed the code, PTAL, thanks.

Reply:
Thanks, while I agree fixing PathExists is the correct thing to do, I think we need to investigate and update all the callers to due to the issue I pointed above.

Reply:
Please see my comments here https://github.com/kubernetes/kubernetes/pull/75645#issuecomment-488757177

Reply:
@jingxu97 I see PR: https://github.com/kubernetes/kubernetes/pull/75645 is fixing all the pathExists issue, shall I make this PR code change only in flexvolume, I would like to cherry-pick to old release, would like to make it as small as possible. Thanks.

Reply:
ping @jingxu97 @msau42 PTAL, thanks.